neonvm/controllers/virtualmachinemigration_controller.go

```diff
@@ -218,8 +218,24 @@ func (r *VirtualMachineMigrationReconciler) Reconcile(ctx context.Context, req c
 		targetRunner := &corev1.Pod{}
 		err := r.Get(ctx, types.NamespacedName{Name: migration.Status.TargetPodName, Namespace: vm.Namespace}, targetRunner)
 		if err != nil && apierrors.IsNotFound(err) {
+			// We require the SSH secret to exist because we cannot unmount and
+			// mount the new secret into the VM after the live migration. If a
+			// VM's SSH secret is deleted accidentally then live migration is
+			// not possible.
+			if len(vm.Status.SSHSecretName) == 0 {
```

Comment:
Maybe you can abstract away this code, because you have very similar code in `doReconcile`. Not sure, maybe there are subtle differences.

Reply:
In the other controller we generate a ssh secret name but here we return an error.

Reply:
secrets mounted to VM as disks (qcow2) ? then they will be transferred to target QEMU on live migration as other block devices

Reply:
Yes, we mount it as a disk (iso not qcow2). Code is here: https://github.com/neondatabase/autoscaling/blob/cad4d78e271c58aa5fad51715780e0b707d6c83a/neonvm/runner/main.go#L589

Reply:
> secrets mounted to VM as disks (qcow2) ? then they will be transferred to target QEMU on live migration as other block devices

Yes, but k8s secret is still needed to put private key on destination.

Reply:
I think actually we don't need to check the secret here.

IIUC: We only get to `vmv1.VmmPending` iff the VM is set to `vmv1.VmPreMigrating`, which can only happen if it first goes through `vmv1.VmRunning`, which requires the secret to be created.

There's an open question about what we should do for old VMs, that don't have an ssh secret created because they're already in `VmRunning`. IMO the correct way to handle that is to have a default-false flag for whether to enable SSH, with `SSHSecretName` getting generated iff it's empty and ssh is enabled. This would unfortunately require _also_ modifying the control plane as part of the rollout here, but I think the operational overhead is worth it for the clarity around how old VMs are handled.

Reply:
> IIUC: We only get to vmv1.VmmPending iff the VM is set to vmv1.VmPreMigrating, which can only happen if it first goes through vmv1.VmRunning, which requires the secret to be created.

Right, but as I wrote in the comment above the secret might get deleted accidentally (for example, by human). So that's why I have the check here.

> There's an open question about what we should do for old VMs, that don't have an ssh secret created because they're already in VmRunning. IMO the correct way to handle that is to have a default-false flag for whether to enable SSH, with SSHSecretName getting generated iff it's empty and ssh is enabled. This would unfortunately require also modifying the control plane as part of the rollout here, but I think the operational overhead is worth it for the clarity around how old VMs are handled.

Good point about old VMs that don't have SSH secret. We can make an assumption here that if the `sshSecretName` is empty then the VM doesn't have an SSH secret and we should generate one here. I didn't get your point about control plane involvement.

Reply:
> secret might get deleted accidentally

IIRC we can prevent that from happening though — I'm not sure if the owner ref alone does it, but a finalizer certainly would.
In any case, I don't think we should accommodate that; it's highly unlikely that someone will run `kubectl delete secret` in prod unless they really know what they're doing (most of that stuff is automated by helm!)

Reply:
> We can make an assumption here that if the sshSecretName is empty then the VM doesn't have an SSH secret and we should generate one here. I didn't get your point about control plane involvement.

basically: VMs _can't_ support ssh access unless they have the vm-builder update, right? So there's no benefit from adding complexity to regenerate ssh keys for VMs that don't already have them, because the image won't support it either way.
So, I'm suggesting that we should have another boolean field on the VM spec to enable ssh access, that defaults to false (i.e. all existing VMs don't have them). The downside is that we'd have to _also_ update the control plane to enable this boolean, but IMO that's worth added clarity we get about handling old VMs.

Reply:
> basically: VMs can't support ssh access unless they have the vm-builder update, right? So there's no benefit from adding complexity to regenerate ssh keys for VMs that don't already have them, because the image won't support it either way.

Ah right, forgot about vm-builder image.

> So, I'm suggesting that we should have another boolean field on the VM spec to enable ssh access, that defaults to false (i.e. all existing VMs don't have them). The downside is that we'd have to also update the control plane to enable this boolean, but IMO that's worth added clarity we get about handling old VMs.

Makes sense. Nothing better comes to my mind.

Reply:
pasting something else from DMs: I changed my mind - IMO actually it's worth the weirdness to have ssh secret recreation in the migration controller — we do runner pod recreation in both places, and so we should be willing to recreate their dependencies (i.e. the ssh secret) in both places as well.
In a future change, we can unify both into a single function.

Reply:
> I'm suggesting that we should have another boolean field on the VM spec to enable ssh access

Why not treat presence or absence of `SSHSecretName` as an indicator for the ssh access?

Reply:
If we exclusively use presence/absence of `SSHSecretName`, semantics are less clear: If a VM is created with `SSHSecretName`, should it auto-generate a new secret? or should it rely on an existing one. IMO the behavior should not depend on the current cluster state (e.g. whether the secret exists). There's also concerns if many VMs use the same secret — if the secret doesn't already exist, we'd (depending on behavior) end up auto-generating a secret that's only owned by a single one of those VMs.

Reply:
@sharnoff a VM cannot be created with `SSHSecretName` since it's part of the status and readonly for the user.

When we create a runner pod for a VM, we should make it in the same way in both controller (vm and migration). Currently, migration controller requires the VM to have the SSH secret, otherwise, it returns an error. It turned out that this approach doesn't work with the existing VMs. Now, It's better to treat the existing and new VMs the same way when we create pods for them. This helps with being consistent avoid trapping into edge cases.

So I'll go with creating secret in the migration controller.

Reply:
implemented in 896a8b298f257208132f79aee8f23a12b099f68e